DS-3572: Check authorization for a specified user instead of currentUser

pnbecker opened the pull request:
https://jira.duraspace.org/browse/DS-3572

pnbecker committed:
DS-3572: Check authorization for a specified user instead of currentUser

pnbecker commented:
We should port this to our master branch as well.

tdonohue commented:
@pnbecker : strangely, Travis seems to have failed to kickoff a CI build on this branch/PR.  Could you try and either recreate this PR or push up a minor change (which you can then revert) to kickstart Travis?

pnbecker commented:
@tdonohue I rebased my PR, this should trigger TravisCI. Thank you for notifying.

terrywbrady commented:
@pnbecker , +1 on the code change.  Do you have any advice on testing/verifying this change?

pnbecker commented:
@terrywbrady Probably it would be good to add some unit tests that check this. :roll_eyes:

The only way I know to test this, is the unit test I've written against DS-3431.

pnbecker committed:
DS-3572: Adding simple unit test for DS-3572.

pnbecker commented:
The unit test is not working as expected, I mark that as WIP.

tomdesair commented on dspace-api/src/main/java/org/dspace/authorize/AuthorizeServiceImpl.java:
```diff
@@ -309,7 +309,7 @@ protected boolean authorize(Context c, DSpaceObject o, int action, EPerson e, bo
 
                 if ((rp.getGroup() != null)
-                        && (groupService.isMember(c, rp.getGroup())))
+                        && (groupService.isMember(c, e, rp.getGroup())))
```
We would also need to add the `userToCheck` variable to the `isAdmin(c, adminObject)` call on line 257

pnbecker replied:
good catch! thanks.

pnbecker replied:
should be solved now.

tomdesair commented on dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java:
```diff
@@ -158,35 +158,32 @@ public boolean isMember(Group owningGroup, Group childGroup) {
-                }
-            }
-            //lookup eperson in normal groups and subgroups
-            return epersonInGroup(context, groupName, currentUser);
```
I would leave the `epersonInGroup` method in place as this was written especially to make the `isMember` method more performant (see ticket https://jira.duraspace.org/browse/DS-3004): Saving the retrieval of the Group object from the database and checking membership with 1 query so that it's easily cacheable by Hibernate.

I think it is too expensive to retrieve all the groups of an eperson (multiple times) for each web request if not absolutely necessary (read: when there are no special groups).

pnbecker replied:
I'll take a look next week.

tomdesair reviewed (changes requested):
I've suggested some changes.

That method in the AuthorizeService was introduced by this commit: https://github.com/DSpace/DSpace/commit/9817028ad02eafcfc314fcd11e68c6aadd8a897a

So it seems we can test this by using JSPUI RequestItemServlet. I'm not a JSPUI expert but it seems that we can test this like this:
* Create a private item with a private bitstream both limited to a certain group.
* As an anonymous user, request access to that item AND its bitstreams.
* As an admin (or as a user member of the above group), approve the item request. 

--> I expect that the private bitstream of that item is attached even though they are not available to the anonymous group because the `authorizeService.authorizeActionBoolean(context, null, bitstream, Constants.READ, false)` call will look at the membership of the current user and not of the anonymous user.

pnbecker commented:
As first action on this PR I fixed the test. No it fails when the PR is not in place and succeeds when the PR is.

This PR needs to be rebased. Some caching was added to `org.dspace.eperson.GroupServiceImpl`. The method affected by the merge conflict is exactly the part of this PR that needs to be improved. Nevertheless, as the code in this PR and the current code in our dspace-6_x branch can be part of the soultion, I don't rebase that now, but leave it to when this PR is overhauled.

pnbecker commented:
I rebased this and updated my fix for the issue. The test currently is failing. I think the culprit is `GroupService.isMember(context, eperson, group)` or `GroupDAOImpl.findByNameAndMembership(Context, String, EPerson)`.

@tomdesair Could you please take a look if you see any reason why the test still fails? Are you sure `GroupDAOImpl.findByNameAndMembership(Context, String, EPerson)` works as expected? Can you please give it another review?

pnbecker commented:
Closing this PR in favor for #1767
